### PR TITLE
[WIP] add search modal

### PR DESCRIPTION
**Pull request by sorawee**

### CODE REVIEW

File: example/wescheme-example.js [hunk 1]
```diff
@@ -38,6 +38,7 @@ document.getElementById('mode').onchange = function(e) {
     cm.doc.clearHistory();
 
     const options = {
+      search: document.getElementById('search'),
```
Comment (schanzer) [inline example/wescheme-example.js]: Can the editor create this node programmatically? This should be as easy as possible to integrate into an existing CM installation.

File: src/less/blocks.less [hunk 2]
```diff
@@ -63,4 +63,30 @@
     overflow: hidden;
     clip: rect(1px 1px 1px 1px);
     clip: rect(1px, 1px, 1px, 1px);
-}
\ No newline at end of file
+}
+
+.CodeMirror {
+    z-index: 0;
```
Comment (schanzer) [inline src/less/blocks.less]: Existing CM installations may require CM to have a different z-index. Do we need to force them to zero?

File: src/searchers/usual.js [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+export default class {
+  constructor(blocks) {
+    this.blocks = blocks;
+  }
+
+  getLabel() {
+    return 'Search by usual';
+  }
+
+  find(forward, e) {
```
Comment (schanzer) [inline src/searchers/usual.js]: It looks like your plan is to have a different JS file for each search type. Personally I don't love the idea (why not just put them all in the same file?), but my bigger concern is duplicate code. Won't they ALL have a `find` function, which works exactly the same way? `initSearch` is also going to be nearly identical. The only thing different will be how `this.searchMatches` is constructed.

File: src/ui/Search.js [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+import {UP, DOWN, ESC, ENTER, PGUP, PGDN, F3} from '../keycode';
+
+// TODO: Make sure to bind modal to your appElement (http://reactcommunity.org/react-modal/accessibility/)
+// Modal.setAppElement('#yourAppElement');
+
+export default class extends Component {
+  constructor(props) {
+    super(props);
+
+    this.state = {showSearchModal: false, searchString: '', searchMode: 0};
```
Comment (schanzer) [inline src/ui/Search.js]: `searchString` isn't relevant for all searches.

File: src/ui/Search.js [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+    if (e.keyCode === ENTER || e.keyCode === ESC) { // enter or escape
+      this.setState({showSearchModal: false});
+
+      // NOTE(Oak): getActiveNode uses document.activeElement, but right now
+      // our focus is on the modal, so we use setTimeout to wait for the modal to close
+      // first
+      setTimeout(() => {
+        if (this.state.searchString === '') return;
+        this.props.searchModes[this.state.searchMode].initSearch(this.state.searchString);
+      });
```
Comment (schanzer) [inline src/ui/Search.js]: How long is the timer?

File: src/ui/Search.js [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+      e.preventDefault(); // prevent the browser search
+
+      this.setState({showSearchModal: true});
+      return;
+
+    case PGUP:
+      e.preventDefault(); // we never want pgup and pgdn to actually do pgup and pgdn
+
+      // NOTE: Oak doesn't think the next line is right. We ought to support
+      // searching in non block mode too, but this's the current behavior
```
Comment (schanzer) [inline src/ui/Search.js]: no - we only support searching in blocks mode

File: src/ui/Search.js [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+        this.props.searchModes[this.state.searchMode].find(false, e);
+      }
+      return;
+
+    case PGDN:
+      e.preventDefault(); // we never want pgup and pgdn to actually do pgup and pgdn
+
+      // NOTE: Oak doesn't think the next line is right. We ought to support
+      // searching in non block mode too, but this's the current behavior
+      if (!this.props.blocks.blockMode) return;
```
Comment (schanzer) [inline src/ui/Search.js]: ditto

File: src/ui/Search.js [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+  constructor(props) {
+    super(props);
+
+    this.state = {showSearchModal: false, searchString: '', searchMode: 0};
+    this.searchBoxNode = null;
+  }
+
+  handleKeyModal = e => {
+    if (e.keyCode === ENTER || e.keyCode === ESC) { // enter or escape
+      this.setState({showSearchModal: false});
```
Comment (schanzer) [inline src/ui/Search.js]: We should call into `this.blocks.sayAndForget` to notify the user that the modal has been closed.

File: src/ui/Search.js [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+      // our focus is on the modal, so we use setTimeout to wait for the modal to close
+      // first
+      setTimeout(() => {
+        if (this.state.searchString === '') return;
+        this.props.searchModes[this.state.searchMode].initSearch(this.state.searchString);
+      });
+
+    } else if (e.keyCode === UP) {
+      if (this.state.searchMode > 0) {
+        this.setState((prevState => ({searchMode: prevState.searchMode - 1})));
```
Comment (schanzer) [inline src/ui/Search.js]: we should announce the new search mode to blind users.

File: src/utils.js [hunk 10]
```diff
@@ -0,0 +1,2 @@
+// give (a,b), produce -1 if a<b, +1 if a>b, and 0 if a=b
+export function poscmp(a, b) { return  a.line - b.line || a.ch - b.ch;  }
```
Comment (schanzer) [inline src/utils.js]: good idea! Would you mind also using this inside `ast.js` and `render.js`?

File: example/wescheme-example.js (hunk 1 above)
Comment (sorawee) [reply example/wescheme-example.js]: Totally, will do that.

File: src/less/blocks.less (hunk 2 above)
Comment (sorawee) [reply src/less/blocks.less]: I really couldn't find other way to not make CodeMirror over the modal except this way...

File: src/searchers/usual.js (hunk 3 above)
Comment (sorawee) [reply src/searchers/usual.js]: You are right. The plan right now is to use inheritance to enable the reuse.

File: src/ui/Search.js (hunk 5 above)
Comment (sorawee) [reply src/ui/Search.js]: I intend it to be 0 (which is the same as not specifying). This is a trick to append it to the event queue.

File: src/ui/Search.js (hunk 6 above)
Comment (sorawee) [reply src/ui/Search.js]: Got it.

File: src/less/blocks.less (hunk 2 above)
Comment (schanzer) [reply src/less/blocks.less]: OK, good enough for now. If we find a reason to revisit, maybe we can programmatically compute the z-index of the modal to be cm.zIndex+1. But let's leave that for now.

File: src/ui/Search.js (hunk 5 above)
Comment (schanzer) [reply src/ui/Search.js]: Minor nit: can you explicitly write in the 0?

File: example/wescheme-example.js (hunk 1 above)
Comment (sorawee) [reply example/wescheme-example.js]: Done.

File: src/ui/Search.js (hunk 5 above)
Comment (sorawee) [reply src/ui/Search.js]: Done.

File: src/ui/Search.js (hunk 9 above)
Comment (sorawee) [reply src/ui/Search.js]: I don't think this is needed (in the new UI the equivalent is switching tab). The following is already read:

> Speak (C) category=nav "Entered dialog"
> Speak (Q) category=nav "Dialog Document Tab list with 2 items"
> Speak (Q) category=nav "Search by string"
> Speak (Q) category=nav "Tab Selected 1 of 2"

press right

> Speak (I) category=nav "Search by block"
> Speak (Q) category=nav "Tab Selected 2 of 2"

**Pull request merged**